src/NHibernate/Engine/ForeignKeys.cs

```diff
@@ -178,10 +178,29 @@ public static bool IsNotTransientSlow(string entityName, object entity, ISession
 				return false;
 			}
 
+			var proxy = entity as INHibernateProxy;
+			if (proxy?.HibernateLazyInitializer.IsUninitialized == true)
+			{
+				return false;
+			}
+
 			// let the interceptor inspect the instance to decide
+			var interceptorResult = session.Interceptor.IsTransient(entity);
+			if (interceptorResult.HasValue)
+				return interceptorResult;
+
 			// let the persister inspect the instance to decide
-			return session.Interceptor.IsTransient(entity) ??
-			       session.GetEntityPersister(entityName, entity).IsTransient(entity, session);
+			if (proxy != null)
+			{
+				// The persister only deals with unproxied entities.
+				entity = proxy.HibernateLazyInitializer.GetImplementation();
```

Comment:
I don't think it's a correct change. That leads to proxy initialization isn't it? Transient fast check should not do any such thing - it should return null if fast detection (without db hit) can't be made

Reply:
OK I see.. It's about only initialized case. Well still it looks kinda hacky...

Reply:
This is the way to get the underlying entity, as used [here](https://github.com/nhibernate/nhibernate-core/blob/21846c017658e89af6c2d720b78624abf9acc064/src/NHibernate/Event/Default/DefaultLoadEventListener.cs#L186).
If the persister should not deal with proxies, I do not see any other way to achieve this.

Reply:
And I have just checked, fixing `GetIdentifier` is not enough. `IsTransient` may also call `GetVersion` which would also call a getter for the concrete class. And there, the only way to get it working would be to unproxy the object anyway. (Or the check on version would have to be given up with proxies, but that would change the behavior between proxy case and not-proxy case, which would not be a good move I think.)

Reply:
That actually a good catch.